tests/unit/OccAnalysisSpec.hs

```diff
@@ -0,0 +1,339 @@
+      -- albeit at different iterations of the `i` loop.  To analyze this
+      -- correctly, we would need to know that `j` and `k` may collide across
+      -- `case` arms, though not within an arm.
+      ann `shouldBe` OccInfo (UsageInfo 2 (1,Unbounded))
+    it "does not crash on indexing by state-effect-bound binders" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  with_state (0 @ Fin 10) \\ref."
+        , "    xs.(get ref)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
+    it "assumes state references touch everything" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  with_state (0 @ Fin 10) \\ref."
+        , "    for i:(Fin 3)."
+        , "      xs.(get ref)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Unbounded))
+    it "assumes state references touch everything even if the initializer doesn't" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  for i:(Fin 10)."
+        , "    with_state i \\ref."
+        , "      xs.(get ref)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Unbounded))
+    it "analyzes through accum effects" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  run_accum (AddMonoid Float) \\ref."
+        , "    for i."
+        , "      ref += xs.i"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
+    -- TODO Should probably construct an example of indexing with the ref of a
+    -- run_accum (and, for that matter, with the handle of a run_state or
+    -- run_accum), but I'm not sure how to make either of those well-typed.
+    it "does not crash on indexing by reader-effect-bound binders" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        , "  with_reader (0 @ Fin 10) \\ref."
+        , "    xs.(ask ref)"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
+    it "understands that while loops access repeatedly" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10) => Float = unreachable ()"
+        -- with_state prevents the access from appearing dead
+        , "  with_state 0 \\ref."
+        , "    while \\_."
+        , "      for i."
+        , "        ref := (get ref) + xs.i"
+        , "      False"
+        ]
+      -- TODO Why is this coming up as 0 indexing depth?  Should be 1.
+      ann `shouldBe` OccInfo (UsageInfo 1 (0,Unbounded))
+    it "understands index-defining bindings" do
+      ann <- analyze cfg env
+        [ ":p"
+        , "  xs : (Fin 10 | Fin 3) => Float = unreachable ()"
+        , "  for i."
+        , "    j = Left i"
+        , "    xs.j"
+        ]
+      ann `shouldBe` OccInfo (UsageInfo 1 (1,Bounded 1))
```

Comment:
This is probably unsurprising, since the `j` binding is an atom binding and will get inlined in simplification?

Reply:
Made the test an expression instead (though that is always "unknown" at the moment).